File: leetcode/1334.阈值距离内邻居最少的城市.go
func findTheCity(n int, edges [][]int, distanceThreshold int) int {
    //floyed 算法
    //f[i][j] = min(f[i][k] + f[k][j], f[i][j])
    f := make([][]int, n)
    for i := 0; i < n; i++ {
        f[i] = make([]int, n)
        for j := range f[i] {
            f[i][j] = math.MaxInt / 2
        }
    }
    for _, e := range edges {
        x, y, wt := e[0], e[1], e[2]
        f[x][y], f[y][x] = wt, wt
    }
    for k := 0; k < n; k++ {
        for i := 0; i < n; i++ {
            for j := 0; j < n; j++ {
                f[i][j] = min(f[i][j], f[i][k] + f[k][j])
            }
        }
    }
    idx := -1
    mx := math.MaxInt
    for i := 0; i < n; i ++ {
        cnt := 0
        for j := 0; j < n; j ++ {
            if j != i && f[i][j] <= distanceThreshold {
                cnt += 1
            }
        }
        if cnt <= mx {
            mx = cnt
            idx = i
        }
    }
    return idx
}

File: leetcode/1334.阈值距离内邻居最少的城市.py
import numpy as np
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        fr, to, we = np.split(np.asarray(edges), 3, -1)
        W = np.full((n, n), inf)
        W[fr, to] = we
        W[to, fr] = we
        W[range(n), range(n)] = 0
        print(W)
        for k in range(n): W = np.minimum(W, W[:, [k]] + W[[k], :])

        V = (W <= distanceThreshold).sum(-1)

        i = 0
        for j, m in enumerate(V):
            if m <= n:
                i, n = j, m

        return i

